pkg/gcp/ocpgwdeployer.go

```diff
@@ -0,0 +1,303 @@
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/stringset"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/cloud-prepare/pkg/ocp"
+	"google.golang.org/api/compute/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+)
+
+type ocpGatewayDeployer struct {
+	gcp          *gcpCloud
+	msDeployer   ocp.MachineSetDeployer
+	instanceType string
+	image        string
+}
+
+// NewOcpGatewayDeployer returns a GatewayDeployer capable deploying gateways using OCP
+// If the supplied cloud is not a gcpCloud, an error is returned
+func NewOcpGatewayDeployer(cloud api.Cloud, msDeployer ocp.MachineSetDeployer, instanceType, image string) (api.GatewayDeployer, error) {
+	gcp, ok := cloud.(*gcpCloud)
+	if !ok {
+		return nil, errors.New("the cloud must be GCP")
+	}
+
+	return &ocpGatewayDeployer{
+		gcp:          gcp,
+		msDeployer:   msDeployer,
+		instanceType: instanceType,
+		image:        image,
+	}, nil
+}
+
+func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, reporter api.Reporter) error {
+	var currentGWInstanceList []*compute.Instance
+
+	reporter.Started(messageCreateExtFWRules)
+
+	externalIngress := newExternalFirewallRules(d.gcp.projectID, d.gcp.infraID, input.PublicPorts)
+	if err := d.gcp.openPorts(externalIngress); err != nil {
+		reporter.Failed(err)
+		return err
+	}
+
+	reporter.Succeeded("Opened External ports %q with firewall rule %q on GCP",
+		formatPorts(input.PublicPorts), externalIngress.Name)
+
+	reporter.Started(messageRetrieveZones)
+
+	zones, err := d.gcp.client.ListZones()
+	if err != nil {
+		return reportFailure(reporter, err, "failed to list the zones in the project %q. %v", d.gcp.projectID)
+	}
+
+	reporter.Succeeded(messageRetrievedZones)
+	reporter.Started(messageValidateCurrentGWCount)
+
+	zonesWithSubmarinerGW := stringset.NewSynchronized()
+
+	for _, zone := range zones.Items {
+		region := zone.Region[strings.LastIndex(zone.Region, "/")+1:]
+		if region != d.gcp.region {
+			continue
+		}
+
+		instanceList, err := d.gcp.client.ListInstances(zone.Name)
+		if err != nil {
+			return reportFailure(reporter, err, "failed to list instances in zone %q of project %q", zone.Name, d.gcp.projectID)
+		}
+
+		for _, instance := range instanceList.Items {
+			hasPublicIP, err := d.gcp.client.InstanceHasPublicIP(instance)
+			if err != nil {
+				return reportFailure(reporter, err, "failed to verify if instance %q has public-ip or not in project %q",
+					instance.Name, d.gcp.projectID)
+			}
+
+			if hasPublicIP {
+				for _, tag := range instance.Tags.Items {
+					if tag == submarinerGatewayNodeTag {
+						currentGWInstanceList = append(currentGWInstanceList, instance)
+
+						if !zonesWithSubmarinerGW.Contains(zone.Name) {
+							zonesWithSubmarinerGW.Add(zone.Name)
+						}
+
+						break
+					}
+				}
+			}
+		}
+	}
+
+	if len(currentGWInstanceList) == input.Gateways {
```

Comment:
`currentGWInstanceList` is now redundant with `zonesWithSubmarinerGW`

```suggestion
	if len(zonesWithSubmarinerGW) == input.Gateways {
```

Reply:
I'm sure you meant zonesWithSubmarinerGW.Size() :)

Reply:
yes 😄